### ISSUE TITLE
Selector Inference: exclude generated CSS module class names

### ISSUE THREAD

**Issue by twschiller** [pixiebrix/pixiebrix-extension#2498]
Context
- CSS module frameworks generate CSS class names (sometimes called name mangling)
- These classnames change each time the style changes
- Sometimes these have a fixed prefix, e.g., `foo-AbCeF`

Implementation Sketch
- Look for seemingly random letters/digits as part of the class name
- Exclude those from the list of suggested selectors

Examples:
- Names generated from our own code: `v3e7tSx46hdIukz0wxkT`

Prior Art:
- https://github.com/Webfit-project/random_string_detection/blob/master/lib/detector.js

**Comment by twschiller** [comment]
Let's try this approach quickly (we can put behind a skunkworks flag):
- Split string by "-" and "_"
- For each part, compute likelihood that it's random (let's vendor the library above as a first try?)
- Find the longest part that's not random
  - If it's the whole string, use that as the selector
  - Otherwise use the [startswith selector](https://www.w3schools.com/cssref/sel_attr_begin.asp). Including the "-" and "_" up until where the first random part occurs

I think we need to fork and submit a PR to css-selector-generator library to provide a custom filter function for classes. (Currently it only support Regex). See https://github.com/fczbkk/css-selector-generator/blob/master/src/selector-class.ts#L9

**Comment by fregante** [comment]
> to provide a custom filter function for classes. 

I opened an issue: https://togithub.com/fczbkk/css-selector-generator/issues/212



> Find the longest part that's not random

If I'm understanding this correctly, you plan to target classes like `Nav-af3Fg` and extract the `[class^="Nav-"]` part. But isn't the random part here targeting what we want? I assume that in the source code they have `Nav-header` and `Nav-link` and they'll be turned into `Nav-af3Fg` and `Nav-4aNd5`. Using `[class^="Nav-"]` will target two completely different objects.

Can we get a number of example sites in here? I'm only seeing these in the wild: (Facebook, Twitter, Reddit)

<img width="140" alt="Screen Shot 15" src="https://user-images.githubusercontent.com/1402241/154790124-a7b44d80-863c-4125-90bb-0e578661c583.png">

<img width="256" alt="Screen Shot 14" src="https://user-images.githubusercontent.com/1402241/154790125-ba959bb1-8529-4b14-a1e5-7385f2d3bb37.png">

<img width="270" alt="Screen Shot 16" src="https://user-images.githubusercontent.com/1402241/154790119-aa4320f5-6586-4b22-9a27-e7ae20c962cf.png">

**Comment by twschiller** [comment]
> But isn't the random part here targeting what we want?

The problem with the random part is that they're more likely to break when the site creator update the interface. (My understanding is the characters are a hash). If we were doing one-time data extraction, then it'd be OK

>  Using [class^="Nav-"] will target two completely different objects

Yep, that's definitely a consideration. We still need to apply logic that the selector that's being generated is unique on the page. (Which also applies in the case of using the whole class name)

In these cases, we're going to need to generate better/more robust selectors. E.g., 1) using the `[data-testid]` attribute instead of the class name, 2) using button:contains("Foo"), etc.

**Comment by fregante** [comment]
> > But isn't the random part here targeting what we want?
> 
> The problem with the random part is that they're more likely to break when the site creator update the interface

Sorry about the misunderstanding, I meant: 

- `Nav-` is useless because it's generic and `-o3ks` is useless because it's random – but **_if it weren't random_**, that's the important part we'd _really_ want.


> the selector that's being generated is unique on the page

I'm confused, I think this depends on what we're looking for:

- panels need a unique selector
- menuItem is fine if not better off with multiple similar items, e.g. `<a class="navItem"/> <a class="navItem"/> <a class="navItem"/>`

Taking a step back:

- if a site uses random classes like the examples I posted it will be IMHO next to impossible to locate an element permanently unless we get lucky with alternative attributes like `aria-label` and such
- locating fields, links and tables may be a little easier since they are rarer elements and maybe repeated over the page

**Comment by twschiller** [comment]
> Nav- is useless because it's generic

To make sure we're on the same page. This PR is about selectors, e.g., for reading data via JQuery. It's not about what we consider for inferring button/panel structure

> Nav- is useless because it's generic

I'm not sure I understand. If there's only one `Nav-` on the page, then that's very useful. I use this approach frequently by manually writing the selector

> -o3ks is useless because it's random – but if it weren't random

With this PR we're trying to apply a heuristic to determine if it's random. For example, if you see a mix of capital letters, lowercase letters, and numbers, it's most likely random. The goal is to see how well in practice a simple heuristic can work

> I'm confused, I think this depends on what we're looking for:

The point of this ticket is to improve our generation of picking a unique element on the page that is more robust to new versions of the front-end

You are correct that menu item can have multiple `class="navItem"` underneath it. But the menu itself we generally want it to be unique unless. The way to get that is by using has `foo:has(> .navItem)`, where the inside of the has can match multiple things on the page

> it will be IMHO next to impossible to locate an element permanently unless we get lucky with alternative attributes like aria-label and such

Let's find out! In my experience helping people use PixieBrix, I'm generally able to manually write a selector. We will of course also have additional work expand the set of patterns we support. E.g., using captions, labels, etc. That is going to be in a separate ticket/PR

> locating fields, links and tables may be a little easier

They are also often easier because they have practical fields (e.g., name, id, a11y tags) in order to function

---

TL;DR: the logic we're implementing is what people manually have to do today to write robust selectors. (See #2: https://medium.com/brick-by-brick/7-bite-sized-tips-for-reliable-web-automation-and-scraping-selectors-2612bc4de2a1). We're going to automate the best practices

**Comment by fregante** [comment]
Can I have some websites to test this on?

**Comment by fregante** [comment]
I think we effectively can only use the detection as a parameter when sorting classes and not to actually detect classes, especially when they're short: 
- `.btn-one` is virtually identical to `.wkr_aba`
- `.Pag-item` is virtually identical to `.zYziq-ix`
- `.title_name--short-2` is virtually identical to `.ti--tl_e2`

I think complete exclusion of strings can only be applied when there's a reasonable confidence that they're random, for example if they're long and there's more than 2 numbers placed in the middle of the string — which would exclude your suggested string for example: `v3e7tSx46hdIukz0wxkT`.

I think it would be way more effective to let the user exclude classes in the rest of the cases because we're much better at this.

**Comment by twschiller** [comment]
> I think it would be way more effective to let the user exclude classes in the rest of the cases because we're much better at this.

That's true for people with a web development background, but not for the target persona in general (including beginner RPA developers)

> when sorting classes and not to actually detect classes

I agree sorting is a good way to go. However, we have to consider the selector generator algorithm -- even if the algorithm finds a unique class name, it still has to consider other class names/selector combinations (because the long unique class name might be a random CSS module hash/name). If it doesn't generate the alternative selectors, we can't penalize the ones that look random

**Comment by fregante** [comment]
> > when sorting classes and not to actually detect classes
> 
> I agree sorting is a good way to go since randomness detection is not exact.

So if this issue moved from exclusion to sorting, we're already doing that. Does it need further logic?

> I need to look into css-selector-generator a bit more. How does it know which classname based selectors to consider (and when to stop/terminate?)

Our problem generally is that we can't hook into `css-selector-generator` **preferences** and change them either way, it's either **include** or **exclude**.

From what I can tell, it just generates a list of selector from every class:

https://github.com/fczbkk/css-selector-generator/blob/7f8a0fdcfadb63d2d646a1f9dbd5794f66956f6a/src/selector-class.ts#L9

 and then loops them (in order) until it finds a class that selects just the requested element:

https://github.com/fczbkk/css-selector-generator/blob/7f8a0fdcfadb63d2d646a1f9dbd5794f66956f6a/src/utilities-selectors.ts#L183

If that doesn't exist, it falls back to nth-child:

https://github.com/fczbkk/css-selector-generator/blob/7f8a0fdcfadb63d2d646a1f9dbd5794f66956f6a/src/index.ts#L60


There's no special preference with the exception of some Angular classes that are hardcoded in the blacklist.

**Comment by fregante** [comment]
It might be best to reopen this issue as something more specific

**Comment by twschiller** [comment]
> So if this issue moved from exclusion to sorting, we're already doing that. Does it need further logic? It might be best to reopen this issue as something more specific

Agree. I'm think what we'll want to do potentially is warn the user if they're using structural selectors / selectors that aren't likely to be reliable

> There's no special preference with the exception of some Angular classes that are hardcoded in the blacklist.

Yep, we're likely going to have to create our own selector generator library that mimics human intuition better
